solr/core/src/java/org/apache/solr/search/combining/ReciprocalRankFusion.java

```diff
@@ -0,0 +1,208 @@
+ * and development in information retrieval (2009)<br>
+ */
+public class ReciprocalRankFusion extends QueriesCombiner {
+  int k;
+
+  public ReciprocalRankFusion(SolrParams requestParams) {
+    super(requestParams);
+    this.k =
+        requestParams.getInt(CombinerParams.COMBINER_RRF_K, CombinerParams.COMBINER_RRF_K_DEFAULT);
+  }
+
+  @Override
+  public QueryResult combine(QueryResult[] rankedLists) {
+    QueryResult combinedRankedList = initCombinedResult(rankedLists);
+    List<DocList> docLists = new ArrayList<>(rankedLists.length);
+    for (QueryResult rankedList : rankedLists) {
+      docLists.add(rankedList.getDocList());
+    }
+    combineResults(combinedRankedList, docLists, false);
+    return combinedRankedList;
+  }
+
+  private Map<Integer, Integer[]> combineResults(
+      QueryResult combinedRankedList,
+      List<DocList> rankedLists,
+      boolean saveRankPositionsForExplain) {
+    Map<Integer, Integer[]> docIdToRanks = null;
+    HashMap<Integer, Float> docIdToScore = new HashMap<>();
+    for (DocList rankedList : rankedLists) {
+      DocIterator docs = rankedList.iterator();
+      int ranking = 1;
+      while (docs.hasNext() && ranking <= upTo) {
+        int docId = docs.nextDoc();
+        float rrfScore = 1f / (k + ranking);
+        docIdToScore.compute(docId, (id, score) -> (score == null) ? rrfScore : score + rrfScore);
+        ranking++;
+      }
+    }
+    Stream<Map.Entry<Integer, Float>> sortedByScoreDescending =
+        docIdToScore.entrySet().stream()
+            .sorted(Collections.reverseOrder(Map.Entry.comparingByValue()));
+
+    int combinedResultsLength = docIdToScore.size();
+    int[] combinedResultsDocIds = new int[combinedResultsLength];
+    float[] combinedResultScores = new float[combinedResultsLength];
+
+    int i = 0;
+    for (Map.Entry<Integer, Float> scoredDoc :
+        sortedByScoreDescending.collect(Collectors.toList())) {
+      combinedResultsDocIds[i] = scoredDoc.getKey();
+      combinedResultScores[i] = scoredDoc.getValue();
+      i++;
+    }
+
+    if (saveRankPositionsForExplain) {
+      docIdToRanks = getRanks(rankedLists, combinedResultsDocIds);
+    }
+
+    DocSlice combinedResultSlice =
+        new DocSlice(
+            0,
+            combinedResultsLength,
+            combinedResultsDocIds,
+            combinedResultScores,
+            combinedResultsLength,
+            combinedResultScores[0],
+            GREATER_THAN_OR_EQUAL_TO);
+    combinedRankedList.setDocList(combinedResultSlice);
+    SortedIntDocSet docSet = new SortedIntDocSet(combinedResultsDocIds, combinedResultsLength);
+    combinedRankedList.setDocSet(docSet);
+
+    return docIdToRanks;
+  }
+
+  private Map<Integer, Integer[]> getRanks(List<DocList> docLists, int[] combinedResultsDocIDs) {
+    Map<Integer, Integer[]> docIdToRanks;
+    docIdToRanks = new HashMap<>();
+    for (int docID : combinedResultsDocIDs) {
+      docIdToRanks.put(docID, new Integer[docLists.size()]);
```

Comment:
Any particular reason for not using Integer.valueOf()?

Reply:
To do what? Isn't Integer.valueOf(int a) used to wrap an integer?
Can you elaborate?

Reply:
oh dear, I completely missed the[] and thought you are create a new Integer by doing: new Integer() and therefore preventing them from being cached. sorry, my bad!